packages/ragbits-core/src/ragbits/core/pipelines/pipeline.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+from types import UnionType
+from typing import Generic, TypeVar, Union, cast, get_type_hints
+
+from generics import get_filled_type
+from typing_utils import Type, issubtype
+
+InputT = TypeVar("InputT")
+OutputT = TypeVar("OutputT")
+
+
+class Step(ABC, Generic[InputT, OutputT]):
+    """
+    A step in a pipeline that processes input data and returns output data
+    """
+
+    @abstractmethod
+    async def run(self, /, input_data: InputT) -> OutputT:
+        """
+        Run the step with the given input data
+        """
+
+
+class Pipeline(Step[InputT, OutputT]):
+    """
+    A pipeline that consists of multiple steps that process input data and return output data
+    """
+
+    def __init__(self, *steps: Step) -> None:
+        self.steps = steps
+        self._validate_intrastep_types()
```

Comment:
Note: during initialization we validate that individual steps have types that are compatible with each other (i.e, that the return type of each step is within bounds of the input type of its following step), **but** due to specifics of the Python generics system, we can validate that the steps have types compatible with the pipeline itself only when the pipeline is run.